File: 54-spiral-matrix/ken/54.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        
        if len(matrix) == 0:
            return []
        
        ans = []
        m = len(matrix)
        n = len(matrix[0])
        
        def runSpiral(startx, starty, endx, endy):
            if len(ans) == m * n:
                return
            
            for i in xrange(starty, endy):
                ans.append(matrix[startx][i])

            for i in xrange(startx + 1, endx):
                ans.append(matrix[i][endy - 1])

            if endx - 1 > startx:
                for i in xrange(endy - 2, starty - 1, -1):
                    ans.append(matrix[endx - 1][i])

            if starty < endy - 1:
                for i in xrange(endx - 2, startx, - 1):
                    ans.append(matrix[i][starty])
            
            runSpiral(startx + 1, starty + 1, endx - 1, endy - 1)
        
        runSpiral(0, 0, m, n)
        
        return ans
    

File: 54-spiral-matrix/leochang/54.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if (matrix.empty()) return {};
        int m = matrix.size();
        int n = matrix[0].size();
        vector<int> ans(m*n);
        
        int col_left_bound = 0, row_top_bound = 0;
        int col_right_bound = n-1, row_bottom_bound = m-1;

        for (int c = 0; c < m*n;) {
            for (int i = col_left_bound; i <= col_right_bound; i++) {
                ans[c++] = matrix[row_top_bound][i];
            }
            row_top_bound++;
            for (int i = row_top_bound; i <= row_bottom_bound; i++) {
                ans[c++] = matrix[i][col_right_bound];
            }
            col_right_bound--;
            if (c == m*n) {
                break;
            }
            for (int i = col_right_bound; i >= col_left_bound; i--) {
                ans[c++] = matrix[row_bottom_bound][i];
            }
            row_bottom_bound--;
            for (int i = row_bottom_bound; i >= row_top_bound; i--) {
                ans[c++] = matrix[i][col_left_bound];
            }
            col_left_bound++;
        }
        return ans;
    }
};
